### PR TITLE
[SYCL][FPGA] Add support for loop count attribute.

### PR DESCRIPTION
**Pull request by zahiraam**
This patch adds support of loop_min, loop_max and loop_avg attributes. They  specify the minimum, maximum, or average number of iterations of a for loop.

### CODE REVIEW

File: clang/test/SemaSYCL/intel-fpga-loops.cpp [hunk 1]
```diff
@@ -66,6 +68,11 @@ void foo_deprecated() {
   // expected-note@+1 {{did you mean to use 'intel::speculated_iterations' instead?}}
   [[intelfpga::speculated_iterations(6)]] for (int i = 0; i != 10; ++i)
       a[i] = 0;
+
+  // expected-warning@+2 {{attribute 'intelfpga::loop_control_avg' is deprecated}}
+  // expected-note@+1 {{did you mean to use 'intel::loop_control_avg' instead?}}
+  [[intelfpga::loop_control_avg(6)]] for (int i = 0; i != 10; ++i)
+      a[i] = 0;
```
Comment (smanna12) [inline clang/test/SemaSYCL/intel-fpga-loops.cpp:75]: This is not needed since this is new attribute and we do not use any intelfpga spelling

File: clang/test/SemaSYCL/intel-fpga-loops.cpp [hunk 2]
```diff
@@ -304,6 +323,11 @@ void zoo() {
   // expected-error@+1 {{duplicate Intel FPGA loop attribute 'nofusion'}}
   [[intel::nofusion]] for (int i = 0; i != 10; ++i)
       a[i] = 0;
+
+  [[intel::loop_control_avg(2)]]
+  // expected-error@+1{{duplicate Intel FPGA loop attribute 'loop_control_avg'}}
+  [[intel::loop_control_avg(2)]] for (int i = 0; i != 10; ++i)
+      a[i] = 0;
```
Comment (smanna12) [inline clang/test/SemaSYCL/intel-fpga-loops.cpp:330]: Could you add a test with disbale_loop_pipeline since no diagnostic is expected i believe.

File: clang/include/clang/Basic/Attr.td [hunk 3]
```diff
@@ -1957,6 +1957,16 @@ def : MutualExclusions<[SYCLIntelFPGAIVDep,
 def : MutualExclusions<[SYCLIntelFPGAMaxConcurrency,
                         SYCLIntelFPGADisableLoopPipelining]>;
 
+def SYCLIntelFPGALoopCountAvg : StmtAttr {
+  let Spellings = [CXX11<"intel","loop_count_avg">];
```
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:1961]:
```suggestion
  let Spellings = [CXX11<"intel", "loop_count_avg">];
```

File: clang/include/clang/Basic/Attr.td [hunk 4]
```diff
@@ -1957,6 +1957,16 @@ def : MutualExclusions<[SYCLIntelFPGAIVDep,
...
+  let Subjects = SubjectList<[ForStmt, CXXForRangeStmt, WhileStmt, DoStmt],
+                             ErrorDiag, "'for', 'while', and 'do' statements">;
+  let Args = [ExprArgument<"NTripCount">];
+  let LangOpts = [SYCLIsDevice, SilentlyIgnoreSYCLIsHost];
+  let HasCustomTypeTransform = 1;
```
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:1966]: Is this needed? The comments on that flag say it applies to type attributes but this is a statement attribute.

File: clang/include/clang/Basic/AttrDocs.td [hunk 5]
```diff
@@ -2950,6 +2950,29 @@ max_concurrency, initiation_interval, or ivdep.
   }];
 }
 
+def SYCLIntelFPGALoopCountAvgAttrDocs : Documentation {
+  let Category = DocCatVariable;
+  let Heading = "intel::loo_control_avg";
```
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:2955]:
```suggestion
  let Heading = "intel::loop_count_avg";
```

File: clang/include/clang/Basic/AttrDocs.td [hunk 6]
```diff
@@ -2950,6 +2950,29 @@ max_concurrency, initiation_interval, or ivdep.
...
+  let Content = [{
+This attribute applies to a loop. Its purpose is to pass to the optimizer the
+average trip count of a loop. This will allow the optimizer to figure out if
```
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:2958]: Do I understand properly that this attribute exists because the optimizer isn't able to figure out the average trip count of a loop and so the hope is that the user will know the answer? Isn't this what profile-guided optimizations are for? (IOW, does this attribute really need to exist?)

File: clang/include/clang/Sema/Sema.h [hunk 7]
```diff
@@ -13419,7 +13419,8 @@ FPGALoopAttrT *Sema::BuildSYCLIntelFPGALoopAttr(const AttributeCommonInfo &A,
                A.getParsedKind() ==
                    ParsedAttr::AT_SYCLIntelFPGAMaxInterleaving ||
                A.getParsedKind() ==
-                   ParsedAttr::AT_SYCLIntelFPGASpeculatedIterations) {
+                   ParsedAttr::AT_SYCLIntelFPGASpeculatedIterations ||
+               A.getParsedKind() == ParsedAttr::AT_SYCLIntelFPGALoopCountAvg) {
```
Comment (AaronBallman) [inline clang/include/clang/Sema/Sema.h:13423]: Nothing for you to fix in your patch, but this is extending another anti-pattern where we have a templated function that still needs to look at the parse kind specifically (so the function looks generic but is not particularly generic).

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 8]
```diff
@@ -558,6 +558,7 @@ static void CheckForIncompatibleSYCLLoopAttributes(
                                                                          Attrs);
   CheckForDuplicationSYCLLoopAttribute<SYCLIntelFPGASpeculatedIterationsAttr>(
       S, Attrs);
+  CheckForDuplicationSYCLLoopAttribute<SYCLIntelFPGALoopCountAvgAttr>(S, Attrs);
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaStmtAttr.cpp:561]: We should find out from the attribute authors why they want this sort of check. FWIW, C++ and C both removed such requirements for standard attributes because this sort of check turned out to be harmful in practice.

(I'm fine leaving the check in this patch, my comment applies generally to all of our custom attributes.)

File: clang/include/clang/Basic/AttrDocs.td (hunk 6 above)
Comment (zahiraam) [reply clang/include/clang/Basic/AttrDocs.td]: I have updated the doc for the attribute. That might be a little more clear now?

**Comment by zahiraam** [comment]
I need to write the loop_count_min and loop_count_max too. I am trying to find out if there is a way to represent this attribute as a single one, loop_count with option min, max, avg instead of having an entry for each one.

**Comment by smanna12** [comment]
> I need to write the loop_count_min and loop_count_max too. I am trying to find out if there is a way to represent this attribute as a single one, loop_count with option min, max, avg instead of having an entry for each one.

I think you can use Accessors  here like
 let Accessors =
    [Accessor<"isMin", [CXX11<"intel", "loop_count_Min">]>];
}


@AaronBallman, what do you think?

**Comment by zahiraam** [comment]
> > I need to write the loop_count_min and loop_count_max too. I am trying to find out if there is a way to represent this attribute as a single one, loop_count with option min, max, avg instead of having an entry for each one.
> 
> I think you can use Accessors here like
> let Accessors =
> [Accessor<"isMin", [CXX11<"intel", "loop_count_min">]>];
> 
> Same for Max or Avg
> 
> This was you can represent this attribute as a single one.
> }
> 
> @AaronBallman, what is your opinion?

It looks like it is the way to go.

**Comment by AaronBallman** [comment]
> > > I need to write the loop_count_min and loop_count_max too. I am trying to find out if there is a way to represent this attribute as a single one, loop_count with option min, max, avg instead of having an entry for each one.
> > 
> > 
> > I think you can use Accessors here like
> > let Accessors =
> > [Accessor<"isMin", [CXX11<"intel", "loop_count_min">]>];
> > Same for Max or Avg
> > This was you can represent this attribute as a single one.
> > }
> > @AaronBallman, what is your opinion?
> 
> It looks like it is the way to go.

Yeah, I think that would be the right approach.

File: clang/include/clang/Basic/Attr.td (hunk 4 above)
Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:1966]: I still don't think this is needed -- the only CI failure I saw when this was removed was a test timeout that seems entirely unrelated.

Comment (zahiraam) [reply clang/include/clang/Basic/Attr.td:1989]: Agreed with @Aaron that is actually needed.

**Comment by zahiraam** [comment]
Next phase will be to allow things like this:
[[intel::loop_control_avg(N)]] [[intel::loop_control_min(N1)]] [[intel::loop_control_max(N2)]]for (unsigned int i = 0; i < n; i++)

{   // code}

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 9]
```diff
@@ -243,6 +243,68 @@ static Attr *handleIntelFPGAIVDepAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
...
+      break;
+    case SYCLIntelFPGALoopCountAttr::CountKind::loop_count_avg:
+      AvgCount++;
+      break;
+    }
+    if (MinCount > 1 || MaxCount > 1 || AvgCount >1 )
+    S.Diag(A->getLocation(), diag::err_sycl_loop_attr_duplication) << 1 << A;
+  }
+}
+
```
Comment (smanna12) [inline clang/lib/Sema/SemaStmtAttr.cpp:283]: need test cases for this duplication.

File: clang/test/SemaSYCL/intel-fpga-loops.cpp [hunk 10]
```diff
@@ -196,6 +201,15 @@ void goo() {
...
+
+  [[intel::loop_count_avg(0)]] for (int i = 0; i != 10; ++i)
+      a[i] = 0;
+  // expected-error@+1 {{'loop_count_avg' attribute requires a non-negative integral compile time constant expression}}
+  [[intel::loop_count_avg(-1)]] for (int i = 0; i != 10; ++i)
+      a[i] = 0;
+  // expected-error@+1 {{'loop_count_avg' attribute requires an integer constant}}
+    [[intel::loop_count_avg("abc")]] for (int i = 0; i != 10; ++i)
+      a[i] = 0;
 }
```
Comment (smanna12) [inline clang/test/SemaSYCL/intel-fpga-loops.cpp:213]: it would be good to add test cases for max and min also

File: clang/test/SemaSYCL/intel-fpga-loops.cpp [hunk 11]
```diff
@@ -402,6 +426,22 @@ void max_concurrency_dependent() {
...
+  for (int i = 0; i != 10; ++i)
+      a[i] = 0;
+
+  [[intel::loop_count_avg(A)]]
+  //expected-error@+1{{duplicate Intel FPGA loop attribute 'loop_count_avg'}}
+  [[intel::loop_count_avg(B)]] for (int i = 0; i != 10; ++i)
+      a[i] = 0;
+
+}
+
```
Comment (smanna12) [inline clang/test/SemaSYCL/intel-fpga-loops.cpp:444]: same here

File: clang/test/SemaSYCL/intel-fpga-loops.cpp [hunk 12]
```diff
@@ -342,6 +361,11 @@ void loop_attrs_compatibility() {
   [[intel::disable_loop_pipelining]]
   [[intel::nofusion]] for (int i = 0; i != 10; ++i)
       a[i] = 0;
+  // no diagnostics are expected
+  [[intel::disable_loop_pipelining]]
+  [[intel::loop_count_avg(8)]]
+  for (int i = 0; i != 10; ++i)
+      a[i] = 0;
```
Comment (smanna12) [inline clang/test/SemaSYCL/intel-fpga-loops.cpp:368]: same here

File: clang/lib/Sema/SemaStmtAttr.cpp (hunk 9 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaStmtAttr.cpp]: Line #438 of SemaSYCL/intel-fpga-loops.cpp acutally pass by this diag here. I am using the same diagnostic. May be I shouldn't?

Comment (smanna12) [reply clang/lib/Sema/SemaStmtAttr.cpp]: Same diagnostic is fine here.

**Comment by zahiraam** [comment]
Review please?

**Review by smanna12 — DISMISSED** [review]
LGTM. @AaronBallman, could you please take a look at the support? Thanks.

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 13]
```diff
@@ -243,6 +243,68 @@ static Attr *handleIntelFPGAIVDepAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
...
+    if (!ArgVal) {
+      S.Diag(E->getExprLoc(), diag::err_attribute_argument_type)
+          << A.getAttrName() << AANT_ArgumentIntegerConstant
+          << E->getSourceRange();
+      return nullptr;
+    }
+
+    if (ArgVal->getSExtValue() < 0) {
+      S.Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << A.getAttrName() << /* non-negative */ 1;
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaStmtAttr.cpp:300]:
```suggestion
          << A << /* non-negative */ 1;
```

**Review by AaronBallman — COMMENTED** [review]
Can you update the review title and description (this adds multiple attributes, none of which are named loop_count)?

File: clang/include/clang/Basic/AttrDocs.td [hunk 14]
```diff
@@ -2955,29 +2955,30 @@ def SYCLIntelFPGALoopCountAttrDocs : Documentation {
...
+used by some of the loop optimization to make decisions such as if the loop
+should be unrolled. It is a way for the user to provide some information without
+using PGO.
 
 .. code-block:: c++
 
   void foo() {
     int a[10];
-    [[intel::loop_count_min(40)] for (int i = 0; i < 10; ++i) a[i] = 0;
+    [[intel::loop_count_min(4)] for (int i = 0; i < n; ++i) a[i] = 0;
```
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:2968]: How about:
```
void foo(int *array, size_t n) {
  [[intel::loop_count_min(4) for (int i = 0; i < n; ++i) array[i] = 0;
}
```
(and similar for the other examples)? That way the examples are compilable.

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 15]
```diff
@@ -605,13 +605,11 @@ MDNode *LoopInfo::createMetadata(
...
-              ConstantInt::get(llvm::Type::getInt32Ty(Ctx),
-                               Attrs.SYCLIntelFPGALoopCountValue[i]))};
+  if (Attrs.SYCLIntelFPGAVariantCount.size() > 0) {
+    for (auto &VC : Attrs.SYCLIntelFPGAVariantCount) {
+      Metadata *Vals[] = {MDString::get(Ctx, VC.first),
+                          ConstantAsMetadata::get(ConstantInt::get(
+                              llvm::Type::getInt32Ty(Ctx), VC.second))};
       LoopProperties.push_back(MDNode::get(Ctx, Vals));
     }
   }
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CGLoopInfo.cpp]:
```suggestion
  for (auto &VC : Attrs.SYCLIntelFPGAVariantCount) {
    Metadata *Vals[] = {MDString::get(Ctx, VC.first),
                        ConstantAsMetadata::get(ConstantInt::get(
                            llvm::Type::getInt32Ty(Ctx), VC.second))};
    LoopProperties.push_back(MDNode::get(Ctx, Vals));
  }
```

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 16]
```diff
@@ -1050,16 +1043,14 @@ void LoopInfoStack::push(BasicBlock *Header, clang::ASTContext &Ctx,
...
-                                  : "llvm.loop.intel.loopcount_avg";
-      setSYCLIntelFPGALoopCountVariant(var);
+      unsigned int count = IntelFPGALoopCountAvg->getNTripCount()
+                               ->getIntegerConstantExpr(Ctx)
+                               ->getSExtValue();
+      const char *var =
+          IntelFPGALoopCountAvg->isMax()   ? "llvm.loop.intel.loopcount_max"
+          : IntelFPGALoopCountAvg->isMin() ? "llvm.loop.intel.loopcount_min"
+                                           : "llvm.loop.intel.loopcount_avg";
+      setSYCLIntelFPGAVariantCount(var, count);
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CGLoopInfo.cpp]:
```suggestion
      unsigned int Count = IntelFPGALoopCountAvg->getNTripCount()
                               ->getIntegerConstantExpr(Ctx)
                               ->getSExtValue();
      const char *Var =
          IntelFPGALoopCountAvg->isMax()   ? "llvm.loop.intel.loopcount_max"
          : IntelFPGALoopCountAvg->isMin() ? "llvm.loop.intel.loopcount_min"
                                           : "llvm.loop.intel.loopcount_avg";
      setSYCLIntelFPGAVariantCount(Var, Count);
```

File: clang/lib/CodeGen/CGLoopInfo.h [hunk 17]
```diff
@@ -414,19 +409,9 @@ class LoopInfoStack {
...
-  void setSYCLIntelFPGALoopCountValue(unsigned C) {
-    StagedAttrs.SYCLIntelFPGALoopCountValue.push_back(C);
-  }
-
-  /// Set value of loopcount variant for the next loop pushed.
-  void setSYCLIntelFPGALoopCountVariant(const char *var) {
-    StagedAttrs.SYCLIntelFPGALoopCountVariant.push_back(var);
+  /// Set value of variant and loop count for the next loop pushed.
+  void setSYCLIntelFPGAVariantCount(const char *var, unsigned int C) {
+    StagedAttrs.SYCLIntelFPGAVariantCount.push_back({var, C});
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CGLoopInfo.h:414]:
```suggestion
  void setSYCLIntelFPGAVariantCount(const char *Var, unsigned int Count) {
    StagedAttrs.SYCLIntelFPGAVariantCount.push_back({Var, Count});
```

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 18]
```diff
@@ -605,6 +605,14 @@ MDNode *LoopInfo::createMetadata(
...
   }
 
+  if (Attrs.SYCLIntelFPGAVariantCount.size() > 0) {
+    for (auto &VC : Attrs.SYCLIntelFPGAVariantCount) {
+      Metadata *Vals[] = {MDString::get(Ctx, VC.first),
+                          ConstantAsMetadata::get(ConstantInt::get(
+                              llvm::Type::getInt32Ty(Ctx), VC.second))};
+      LoopProperties.push_back(MDNode::get(Ctx, Vals));
+    }
+  }
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CGLoopInfo.cpp:615]:
```suggestion
  for (auto &VC : Attrs.SYCLIntelFPGAVariantCount) {
    Metadata *Vals[] = {MDString::get(Ctx, VC.first),
                        ConstantAsMetadata::get(ConstantInt::get(
                            llvm::Type::getInt32Ty(Ctx), VC.second))};
    LoopProperties.push_back(MDNode::get(Ctx, Vals));
  }
```

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 19]
```diff
@@ -1030,6 +1041,19 @@ void LoopInfoStack::push(BasicBlock *Header, clang::ASTContext &Ctx,
...
+            dyn_cast<SYCLIntelFPGALoopCountAttr>(A)) {
+      unsigned int count = IntelFPGALoopCountAvg->getNTripCount()
+                               ->getIntegerConstantExpr(Ctx)
+                               ->getSExtValue();
+      const char *var = IntelFPGALoopCountAvg->isMax()
+                            ? "llvm.loop.intel.loopcount_max"
+                            : IntelFPGALoopCountAvg->isMin()
+                                  ? "llvm.loop.intel.loopcount_min"
+                                  : "llvm.loop.intel.loopcount_avg";
+      setSYCLIntelFPGAVariantCount(var, count);
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CGLoopInfo.cpp:1054]:
```suggestion
      unsigned int Count = IntelFPGALoopCountAvg->getNTripCount()
                               ->getIntegerConstantExpr(Ctx)
                               ->getSExtValue();
      const char *Var = IntelFPGALoopCountAvg->isMax()
                            ? "llvm.loop.intel.loopcount_max"
                            : IntelFPGALoopCountAvg->isMin()
                                  ? "llvm.loop.intel.loopcount_min"
                                  : "llvm.loop.intel.loopcount_avg";
      setSYCLIntelFPGAVariantCount(Var, Count);
```

File: clang/lib/CodeGen/CGLoopInfo.h [hunk 20]
```diff
@@ -15,6 +15,7 @@
 #define LLVM_CLANG_LIB_CODEGEN_CGLOOPINFO_H
 
 #include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMap.h"
```
Comment (AaronBallman) [inline clang/lib/CodeGen/CGLoopInfo.h:18]: Is this change needed?

File: clang/include/clang/Basic/AttrDocs.td [hunk 21]
```diff
@@ -2951,6 +2951,41 @@ or ``ivdep``.
   }];
 }
 
+def SYCLIntelFPGALoopCountAttrDocs : Documentation {
+  let Category = DocCatVariable;
+  let Heading = "intel::loop_count_min, intel::loop_count_max, intel::loop_count_avg";
+  let Content = [{
+The loop count attribute specifies the minimum, maximum, or average number of
```
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:2958]:
```suggestion
The loop count attributes specify the minimum, maximum, or average number of
```

**Review by AaronBallman — APPROVED** [review]
LGTM aside from a super small grammar nit with the documentation.

File: clang/include/clang/Basic/AttrDocs.td (hunk 21 above)
Comment (zahiraam) [reply clang/include/clang/Basic/AttrDocs.td:2958]: Ouch!

**Review by AaronBallman — DISMISSED** [review]
LGTM!

File: clang/lib/CodeGen/CGLoopInfo.cpp [hunk 22]
```diff
@@ -553,10 +553,10 @@ MDNode *LoopInfo::createMetadata(
...
   // Setting max_concurrency attribute with number of threads
   if (Attrs.SYCLMaxConcurrencyEnable) {
-    Metadata *Vals[] = {MDString::get(Ctx, "llvm.loop.max_concurrency.count"),
-                        ConstantAsMetadata::get(ConstantInt::get(
-                            llvm::Type::getInt32Ty(Ctx),
-                            Attrs.SYCLMaxConcurrencyNThreads))};
+    Metadata *Vals[] = {
+        MDString::get(Ctx, "llvm.loop.max_concurrency.count"),
+        ConstantAsMetadata::get(ConstantInt::get(
+            llvm::Type::getInt32Ty(Ctx), Attrs.SYCLMaxConcurrencyNThreads))};
```
[the remark this answers is not in the document]
Comment (smanna12) [reply clang/lib/CodeGen/CGLoopInfo.cpp:559]: LGTM. one comment- @zahiraam, could you remove this unrelated change?

**Review by smanna12 — APPROVED** [review]

**Review by premanandrao — APPROVED** [review]
LGTM
